### PR TITLE
Transparent inputs

### LINKED ISSUE #10: Add the abililty to use transparent inputs

### PR DESCRIPTION
**Pull request by Duddino**
Adds the ability to make Transaction with transparent inputs.
Mixing inputs is not supported yet.
Closes #10.

### CODE REVIEW

File: js/pivx_shielding.js [hunk 1]
```diff
@@ -137,21 +144,25 @@ export default class PIVXShielding {
   }
 
   /**
-   * Creates a transaction, sending `amount` satoshis to the address
-   * @param {{address: String, amount: Number}} target
+   * Createes a transaction, sending `amount` satoshis to the address
```
Comment (panleone) [inline js/pivx_shielding.js:147]: typo

File: js/pivx_shielding.js [hunk 2]
```diff
@@ -137,21 +144,25 @@ export default class PIVXShielding {
...
+   * @param {{address: String, amount: Number, blockHeight: Number, useShieldInputs: bool}} target
```
Comment (panleone) [inline js/pivx_shielding.js:148]: are you sure that blockHeight is needed?

File: js/pivx_shielding.js [hunk 3]
```diff
@@ -82,10 +84,15 @@ export default class PIVXShielding {
...
-     * @type {Map<String, String[]>} A map txid->nullifiers, storing pending transaction.
+     * @type {Map<String, {nullifiers: String[], isShield: bool}>} A map txid->nullifiers, storing pending transaction.
      * @private
      */
     this.pendingTransactions = new Map();
+
+    /**
+     * @type {{txid: String, vout: Number, amount: Number, privateKey: Uint8Array, script: Uint8Array}[]}
+     */
+    this.utxos = [];
```
Comment (panleone) [inline js/pivx_shielding.js:95]: this is useelss and we can pass utxos directly inside the createTransaction function

File: js/pivx_shielding.js [hunk 4]
```diff
@@ -82,10 +84,15 @@ export default class PIVXShielding {
     this.unspentNotes = [];
 
     /**
-     * @type {Map<String, String[]>} A map txid->nullifiers, storing pending transaction.
+     * @type {Map<String, {nullifiers: String[], isShield: bool}>} A map txid->nullifiers, storing pending transaction.
```
Comment (panleone) [inline js/pivx_shielding.js:87]: again is useless handling normal utxos here

File: js/pivx_shielding.js [hunk 5]
```diff
@@ -163,11 +174,14 @@ export default class PIVXShielding {
...
    */
   finalizeTransaction(txid) {
-    const nullifiers = this.pendingTransactions.get(txid);
-    if (!nullifiers) {
-      throw new Error(`Unknown transaction ${txid}`);
+    const { nullifiers, isShield } = this.pendingTransactions.get(txid);
+    if (isShield) {
+      this.removeSpentNullifiers(nullifiers);
+    } else {
+      this.utxos = this.utxos.filter((u) =>
```
Comment (panleone) [inline js/pivx_shielding.js:181]: again is useless handling utxo logic inside the library

File: src/transaction.rs [hunk 6]
```diff
@@ -159,31 +164,60 @@ pub struct JSTransaction {
...
+        utxos,
+    } = serde_wasm_bindgen::from_value::<JSTxOptions>(options).expect("Cannot deserialize notes");
+
+    let extsk = decode_extsk(&extsk, is_testnet);
     let network = if is_testnet {
         Network::TestNetwork
     } else {
         Network::MainNetwork
     };
+    let input = if let Some(mut notes) = notes {
```
Comment (panleone) [inline src/transaction.rs:207]: add in documentation that notes have priority? or throw error if user insert both utxo and notes?

File: src/transaction.rs [hunk 7]
```diff
@@ -205,11 +239,93 @@ pub fn create_transaction_internal(
...
+
+    prove_transaction(builder, nullifiers)
+}
 
+fn choose_utxos(
+    builder: &mut Builder<Network, OsRng>,
+    utxos: &[Utxo],
+    amount: u64,
+) -> Result<(Vec<String>, Amount), Box<dyn Error>> {
+    let fee = 2365000u64;
```
Comment (panleone) [inline src/transaction.rs:282]: as I said in the last PR this hardcoded fee will create problems, but if you want we can fix in another PR

**Review by panleone — CHANGES_REQUESTED** [review]
First review with logic part that I don't like, will do a second review where I test functionality

File: js/pivx_shielding.js (hunk 2 above)
Comment (Duddino) [reply js/pivx_shielding.js:148]: Builder takes it as a parameter

Comment (panleone) [reply js/pivx_shielding.js:148]: right but I'm sure that we can just pass a symbolic height in which sapling was activated

Comment (Duddino) [reply js/pivx_shielding.js:148]: I think zcash uses the expiry height in their transaction, but PIVX doesn't, so it should in theory work

File: src/transaction.rs (hunk 6 above)
Comment (Duddino) [reply src/transaction.rs:209]: Added an assert!

File: src/transaction.rs (hunk 7 above)
Comment (Duddino) [reply src/transaction.rs:326]: Will fix in another PR

File: js/pivx_shielding.js (hunk 3 above)
Comment (Duddino) [reply js/pivx_shielding.js:104]: After some internal discussion, we've decided to remove this functionality and add it if needed in the future

**Pull request merged**